File: lib/services/cart_service.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/product.dart';
import 'api_service.dart';

class CartItem {
  final int productId;
  final String name;
  final double price;
  int quantity;
  final String? image;

  CartItem({
    required this.productId,
    required this.name,
    required this.price,
    required this.quantity,
    this.image,
  });

  Map<String, dynamic> toJson() => {
        'productId': productId,
        'name': name,
        'price': price,
        'quantity': quantity,
        'image': image,
      };

  factory CartItem.fromJson(Map<String, dynamic> json) => CartItem(
        productId: json['productId'],
        name: json['name'],
        price: (json['price'] as num).toDouble(),
        quantity: json['quantity'],
        image: json['image'],
      );
}

class CartService extends ChangeNotifier {
  List<CartItem> _items = [];
  List<CartItem> get items => _items;
  final ApiService _apiService = ApiService();

  CartService() {
    _loadFromStorage();
  }

  Future<void> _loadFromStorage() async {
    final prefs = await SharedPreferences.getInstance();
    final String? raw = prefs.getString('cart_items');
    if (raw != null) {
      final List<dynamic> list = jsonDecode(raw);
      _items = list.map((i) => CartItem.fromJson(i)).toList();
      notifyListeners();
    }
  }

  Future<void> _saveToStorage() async {
    final prefs = await SharedPreferences.getInstance();
    final String raw = jsonEncode(_items.map((i) => i.toJson()).toList());
    await prefs.setString('cart_items', raw);
  }

  void addToCart(Product product, {int qty = 1}) {
    final idx = _items.indexWhere((i) => i.productId == product.id);
    if (idx == -1) {
      _items.add(CartItem(
        productId: product.id,
        name: product.name,
        price: product.price,
        quantity: qty,
        image: product.image,
      ));
    } else {
      _items[idx].quantity += qty;
    }
    _saveToStorage();
    notifyListeners();
  }

  void removeFromCart(int productId) {
    _items.removeWhere((i) => i.productId == productId);
    _saveToStorage();
    notifyListeners();
  }

  void clear() {
    _items = [];
    _saveToStorage();
    notifyListeners();
  }

  double get totalAmount => _items.fold(0, (sum, item) => sum + (item.price * item.quantity));

  Future<void> placeOrder(Map<String, dynamic> orderPayload) async {
    try {
      final response = await _apiService.post('/orders/place-order', orderPayload);
      // response is already decoded JSON, if we get here without exception, order was placed successfully
    } catch (e) {
      throw Exception('Failed to place order: $e');
    }
  }

  Future<void> checkout() async {
    if (_items.isEmpty) return;

    List<String> errors = [];
    for (var item in _items) {
      try {
        final payload = {
          'orderDate': DateTime.now().toIso8601String(),
          'productId': item.productId,
          'qteOrdered': item.quantity,
        };
        await placeOrder(payload);
      } catch (e) {
        errors.add('Product ${item.name}: ${e.toString()}');
      }
    }

    if (errors.isEmpty) {
      clear();
    } else {
      throw Exception('Some items could not be ordered:\n${errors.join('\n')}');
    }
  }
}

File: src/app/services/cart.service.ts
import { Injectable, Inject } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { APP_CONFIG } from '../../main';

export interface CartItem {
  productId: number;
  name: string;
  price: number;
  quantity: number;
  image?: string | null;
  stock?: number;
}

@Injectable({ providedIn: 'root' })
export class CartService {
  private readonly itemsSubject = new BehaviorSubject<CartItem[]>(this.loadFromStorage());
  readonly items$ = this.itemsSubject.asObservable();

  // discount percent applied to the user (0-100)
  private readonly discountSubject = new BehaviorSubject<number>(this.loadDiscountFromStorage());
  readonly discount$ = this.discountSubject.asObservable();

  private readonly apiUrl: string;

  constructor(private readonly http: HttpClient, @Inject(APP_CONFIG) private readonly config: any) {
    const base = (this.config?.apiUrl || '').replace(/\/$/, '');
    // backend orders endpoint is under /api/orders
    this.apiUrl = base.replace(/\/api$/, '') + '/api/orders';
  }

  private saveDiscountToStorage(pct: number) {
    localStorage.setItem('user_discount_percent', String(pct || 0));
  }

  private loadDiscountFromStorage(): number {
    const raw = localStorage.getItem('user_discount_percent');
    const v = raw ? Number(raw) : 0;
    return isNaN(v) ? 0 : v;
  }

  getDiscountPercent(): number {
    return this.discountSubject.value || 0;
  }

  setDiscountPercent(pct: number) {
    const n = Math.max(0, Math.min(100, Number(pct) || 0));
    this.discountSubject.next(n);
    this.saveDiscountToStorage(n);
  }

  getTotalWithDiscount(items: CartItem[]): number {
    const subtotal = items.reduce((s, it) => s + (it.price * it.quantity), 0);
    const pct = this.getDiscountPercent();
    const discounted = subtotal * (1 - pct / 100);
    return Math.max(0, discounted);
  }

  private saveToStorage(items: CartItem[]) {
    localStorage.setItem('cart_items', JSON.stringify(items));
  }

  private loadFromStorage(): CartItem[] {
    const raw = localStorage.getItem('cart_items');
    return raw ? JSON.parse(raw) : [];
  }

  getItems(): CartItem[] {
    return this.itemsSubject.value;
  }

  addToCartProduct(product: { id: number; name: string; price: number; images?: string[]; stock?: number }, qty = 1) {
    const items = [...this.getItems()];
    const idx = items.findIndex(i => i.productId === product.id);
    const image = product.images?.length ? product.images[0] : null;
    if (idx === -1) {
      items.push({ productId: product.id, name: product.name, price: product.price, quantity: qty, image, stock: product.stock });
    } else {
      items[idx].quantity += qty;
      // update stock if provided
      if (typeof product.stock === 'number') {
        items[idx].stock = product.stock;
      }
    }
    this.itemsSubject.next(items);
    this.saveToStorage(items);
  }

  addItem(item: CartItem) {
    this.addToCartProduct({ id: item.productId, name: item.name, price: item.price, images: item.image ? [item.image] : [], stock: item.stock }, item.quantity);
  }

  removeItem(productId: number) {
    const items = this.getItems().filter(i => i.productId !== productId);
    this.itemsSubject.next(items);
    this.saveToStorage(items);
  }

  clear() {
    this.itemsSubject.next([]);
    this.saveToStorage([]);
  }

  // place order using backend
  placeOrder(orderPayload: any) {
    const url = `${this.apiUrl}/place-order`;
    console.debug('[CartService] placeOrder ->', { url, orderPayload, tokenPreview: (localStorage.getItem('jwt_token') || '').slice(0, 20) + '...' });
    return this.http.post<any>(url, orderPayload);
  }
}
